compiler/ir/backend.common/src/org/jetbrains/kotlin/backend/common/lower/loops/ProgressionHandlers.kt

```diff
@@ -221,6 +217,302 @@ internal class UntilHandler(private val context: CommonBackendContext, private v
     }
 }
 
+/** Builds a [HeaderInfo] for progressions built using the `step` extension function. */
+internal class StepHandler(
+    private val context: CommonBackendContext,
+    private val visitor: IrElementVisitor<HeaderInfo?, Nothing?>
+) : ProgressionHandler {
+
+    private val symbols = context.ir.symbols
+
+    override val matcher = SimpleCalleeMatcher {
+        singleArgumentExtensionBySimpleType(FqName("kotlin.ranges.step"), symbols.progressionClassesTypes)
+        parameterCount { it == 1 }
+        parameter(0) { it.type.isInt() || it.type.isLong() }
+    }
+
+    override fun build(expression: IrCall, data: ProgressionType, scopeOwner: IrSymbol): HeaderInfo? =
+        with(context.createIrBuilder(scopeOwner, expression.startOffset, expression.endOffset)) {
+            // Retrieve the HeaderInfo from the underlying progression (if any).
+            val nestedInfo = expression.extensionReceiver!!.accept(visitor, null) as? ProgressionHeaderInfo
+                ?: return null
+
+            val stepArg = expression.getValueArgument(0)!!
+            // We can return the nested info if its step is constant and its absolute value is the same as the step argument. Examples:
+            //
+            //   1..10 step 1               // Nested step is 1, argument is 1. Equivalent to `1..10`.
+            //   10 downTo 1 step 1         // Nested step is -1, argument is 1. Equivalent to `10 downTo 1`.
+            //   10 downTo 1 step 2 step 2  // Nested step is -2, argument is 2. Equivalent to `10 downTo 1 step 2`.
+            if (stepArg.constLongValue != null && nestedInfo.step.constLongValue?.absoluteValue == stepArg.constLongValue) {
+                return nestedInfo
+            }
+
+            // To reduce local variable usage, we create and use temporary variables only if necessary.
```

Comment:
Does this actually have any effect? At least on the JVM, `JvmBuiltinOptimizationLowering` would remove these variables anyway.

Reply:
It does make a difference. `JvmBuiltinOptimizationLowering` does not:
1. Eliminate the variable or its initializer. Not sure if these get optimized out at the bytecode level later on, I don't believe so.
2. Replace gets of variables that are initialized by gets of variables.

Now these _could_ be addressed in the lowering or some other lowering (e.g., constant propagation or something similar).